RealmSwift/SyncSubscription.swift

```diff
@@ -0,0 +1,250 @@
+    case complete
+    // The Subscription encountered an error.
+    case error(Error)
+    // The server is processing the subscription and updating the Realm data
+    // with new matches
+    case pending
+}
+
+extension SyncSubscriptionState: Equatable {
+    static func == (lhs: SyncSubscriptionState, rhs: SyncSubscriptionState) -> Bool {
+        true
+    }
+}
+
+internal protocol AnySyncSubscription {}
+
+internal class SyncSubscription<Element: Object>: AnySyncSubscription {
+    internal typealias QueryFunction = (Query<Element>) -> Query<Element>
+
+    // When the subscription was created. Recorded automatically.
+    internal var createdAt: Date = Date()
+
+    // When the subscription was last updated. Recorded automatically.
+    internal var updatedAt: Date = Date()
+
+    // Name of the subscription, if not specified it will return the value in Query
+    internal var name: String = ""
+
+    #if swift(>=5.4)
+    // Update query for subscription
+    internal func update(@QueryBuilder _ to: () -> (AnySyncSubscription)) throws {
+        fatalError()
+    }
+    #endif // swift(>=5.4)
+
+    private(set) internal var query: QueryFunction
+    internal init(name: String = "", query: @escaping QueryFunction) {
+        self.name = name
+        self.query = query
+    }
+}
+
+#if swift(>=5.4)
+internal protocol QueryBuilderComponent {}
+
+@resultBuilder internal struct QueryBuilder {
+    internal static func buildBlock(_ components: AnySyncSubscription...) -> [AnySyncSubscription] {
+        return components
+    }
+
+    internal static func buildBlock(_ component: AnySyncSubscription) -> AnySyncSubscription {
+        return component
+    }
+}
+
+internal protocol AnyQueryBuilderComponent {}
+#endif // swift(>=5.4)
+
+// Realm operations
+// Realm will only allow getting all the subscriptions and subscribe to a query
+extension Realm {
+    // Get the active subscription set for this realm.
+    internal var subscriptions: [AnySyncSubscription] {
+        fatalError()
+    }
+}
+
+// TODO: Can we observer changes on the subscription set?
+// Task to get state changes from the write transaction
+// @frozen
+internal struct SyncSubscriptionTask {
+    // Notifies state changes for the write subscription transaction
+    // if state is complete this will return complete, will
+    // throw an error if someone updates the subscription set while waiting
+    internal func observe(_ block: @escaping (SyncSubscriptionState) -> Void) {
+        fatalError()
+    }
+}
+
+#if swift(>=5.5) && canImport(_Concurrency)
+@available(macOS 12.0, tvOS 15.0, iOS 15.0, watchOS 8.0, *)
+extension SyncSubscriptionTask {
+    // Notifies state changes for the write subscription transaction
+    // if state is complete this will return complete, will
+    // throw an error if someone updates the subscription set while waiting
+    internal func observe() -> AsyncStream<SyncSubscriptionState> {
```

Comment:
If we're using an AsyncStream here, this should be a computed variable named `state`.